PLFM-6516: Worker to auto cleanup multipart uploads

marcomarasca opened the pull request.

marcomarasca committed:
PLFM-6516: New method to update an old hash

marcomarasca committed:
PLFM-6516: Make sure the updatedOn is changed when updating the upload status

marcomarasca committed:
PLFM-6516: Change hash instead of deleting multipart on restart

marcomarasca committed:
PLFM-6516: Unused method

marcomarasca committed:
PLFM-6516: Added method to fetch old multipart uploads

marcomarasca committed:
PLFM-6516: Abort multipart implementation

marcomarasca committed:
PLFM-6516: Batch part deletes + integration test

marcomarasca committed:
Reduce the data access notification worker interval

marcomarasca committed:
PLFM-6516: Additional test coverage

marcomarasca committed:
PLFM-6516: Multipart cleanup worker

marcomarasca committed:
PLFM-6516: Random hash on restart

marcomarasca committed:
Merge branch 'develop' into PLFM-6516

marcomarasca committed:
PLFM-6516: Worker integration test

john-hill commented on lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/MultipartUploadDAOImpl.java:
```diff
@@ -109,6 +111,13 @@
+			+ " WHERE " + COL_MULTIPART_STARTED_BY + " = ? AND " + COL_MULTIPART_REQUEST_HASH + " = ?";
+	
+	private static final String SQL_SELECT_BATCH = "SELECT " + COL_MULTIPART_UPLOAD_ID + " FROM " + TABLE_MULTIPART_UPLOAD
+			+ " WHERE " + COL_MULTIPART_UPDATED_ON + " < ? ORDER BY " + COL_MULTIPART_STARTED_ON + " LIMIT ?";
```
order by updated on

john-hill commented on lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/file/MultipartUploadDAOImplTest.java:
```diff
@@ -306,55 +309,72 @@ public void testAddPartToAndErrorToUpload() {
-				"Adding a part must update the etag of the master row.");
+		
+		assertNotEquals(status.getEtag(), updated.getEtag(), "Adding a part must update the etag of the master row.");
+		assertNotEquals(status.getMultipartUploadStatus().getUpdatedOn(), updated.getMultipartUploadStatus().getUpdatedOn());
```
greater than

john-hill commented on services/workers/src/main/java/org/sagebionetworks/file/worker/MultipartCleanupWorker.java:
```diff
@@ -0,0 +1,74 @@
+			return;
+		}
+		
+		Instant modifiedBefore = Instant.now().minus(MultipartManagerV2.EXPIRE_PERIOD);
```
let the db machine decide what time it is.

john-hill commented on lib/lib-upload/src/test/java/org/sagebionetworks/upload/multipart/S3MultipartUploadDAOImplTest.java:
```diff
@@ -678,6 +690,205 @@ public void testCreatePartUploadCopyPresignedUrlWithSourceBucketKey() {
+		
+		doNothing().when(mockS3Client).abortMultipartUpload(any());
+		
+		List<String> partKeys = IntStream.range(1, S3MultipartUploadDAOImpl.S3_BATCH_DELETE_SIZE * 2).boxed().map(n -> String.valueOf(n)).collect(Collectors.toList());
```
if the batch size is an input the test can be smaller and simpler

john-hill commented on lib/lib-upload/src/test/java/org/sagebionetworks/upload/multipart/S3MultipartUploadDAOImplTest.java:
```diff
@@ -678,6 +690,205 @@ public void testCreatePartUploadCopyPresignedUrlWithSourceBucketKey() {
+		dao.abortMultipartRequest(request);
+		
+		verify(mockS3Client).deleteObjects(any());
+		verify(mockS3Client).abortMultipartUpload(any());
```
check that it is logged

marcomarasca committed:
PLFM-6516: Fix wrong order by

marcomarasca committed:
PLFM-6516: Let the DB compute the expire time

marcomarasca committed:
PLFM-6516: Do not fail on exceptions

When we try to abort an old multipart upload we do not want to send the worker in a loop where we keep failing. We just try to abort the uploads once if possible.
